src/lib/utils/ct_utils.h

```diff
@@ -75,6 +75,29 @@ constexpr inline void unpoison(T& p) {
    BOTAN_UNUSED(p);
 }
 
+/**
+* This function returns its argument, but (if called in a non-constexpr context)
+* attemps to prevent the compiler from reasoning about the value or the possible
+* range of values. Such optimizations have a way of breaking constant time code.
+*/
+template <typename T>
+constexpr inline T value_barrier(T x) {
+   if(!std::is_constant_evaluated()) {
+      /*
+      * For compilers without inline asm, is there something else we can do?
+      *
+      * For instance we could potentially launder the value through a
+      * `volatile T` or `volatile T*`. This would require some experimentation.
+      */
+#if defined(BOTAN_USE_GCC_INLINE_ASM)
+      asm("" : "+r"(x) : /* no input */);
```

Comment:
As per [this blog post](https://theunixzoo.co.uk/blog/2021-10-14-preventing-optimisations.html):

```suggestion
      asm volatile("" : "+r,m"(x) : : "memory);
```

... i don't have much experience with such inline assembler kung-fu, though.

Reply:
Here's the original source that the blog post above dissects: https://github.com/google/benchmark/blob/7f0e99af540a333108b92d792923ec7fc9e9fad9/include/benchmark/benchmark.h#L470-L477

... for reasons, they distinguish clang and non-clang.

Reply:
BoringSSL is currently using `"+r"(x)` for their value barrier. And per valgrind it does seem sufficient, at least with current Clang. Clobbering all memory, especially in the very low level loops where this kind of thing is useful, may really degrade optimizations. Contra to a benchmark, where the clobbering happens at the very top of the call stack. I left a comment though that we may need a "stronger" asm statement in the future, along with a link to that blog post.